gax/src/main/java/com/google/api/gax/rpc/ApiExceptionFactory.java

```diff
@@ -29,15 +29,103 @@
-  private ApiExceptionFactory() {}
+
+  private static final ExceptionTransformation DEFAULT_EXCEPTION_TRANSFORMATION =
+      new IdentityExceptionTransformation();
+  private static final ExceptionTransformation transform = loadExceptionTransformation();
 
   public static ApiException createException(
       Throwable cause, StatusCode statusCode, boolean retryable) {
+    ApiException exception = mapStatusToException(cause, statusCode, retryable);
+    exception = transform.transform(exception);
+    return validate(exception, statusCode);
+  }
+
+  public static ApiException createException(
+      String message, Throwable cause, StatusCode statusCode, boolean retryable) {
+    ApiException exception = mapStatusToException(message, cause, statusCode, retryable);
+    exception = transform.transform(exception);
+    return validate(exception, statusCode);
+  }
+
+  /**
+   * Ensures that we will return a valid exception.
+   *
+   * @throws IllegalStateException if the exception is null.
+   */
+  private static ApiException validate(ApiException exception, StatusCode statusCode) {
+    if (exception == null) {
+      throw new IllegalStateException(
+          "Unable to map "
+              + statusCode
+              + " to exception. Possibly a bad ExceptionTransformation service provider?");
+    }
+    return exception;
+  }
+
+  /**
+   * Provides an interface to allow applying an arbitrary transformation to the default exception
+   * mapping.
+   *
+   * <p>This can be useful to map the generic {@link ApiException} implementations to a client
+   * library specific implementation. For instance, to extract server-provided information from the
+   * gRPC trailers and present it to the user.
+   */
+  public interface ExceptionTransformation {
+    /** Transforms a given exception to another exception. */
+    ApiException transform(ApiException exception);
+  }
+
+  /** Default exception transformation which returns the unmodified input. */
+  public static class IdentityExceptionTransformation implements ExceptionTransformation {
+
+    @Override
+    public ApiException transform(ApiException exception) {
+      return exception;
+    }
+  }
+
+  private ApiExceptionFactory() {}
+
+  /**
+   * Loads an implementation of {@link ApiExceptionFactory} via {@link ServiceLoader}. Defers to a
+   * default implementation if no {@link ServiceLoader} is found.
+   *
+   * @return the implementation of {@link ApiExceptionFactory} to use.
+   * @throws IllegalStateException if multiple definitions of {@link ServiceLoader} are found.
+   */
+  private static ExceptionTransformation loadExceptionTransformation() {
+    ServiceLoader<ExceptionTransformation> loader =
+        ServiceLoader.load(ExceptionTransformation.class);
+    Iterator<ExceptionTransformation> it = loader.iterator();
+    if (!it.hasNext()) {
+      return DEFAULT_EXCEPTION_TRANSFORMATION;
+    }
+    ExceptionTransformation transform = it.next();
+    if (it.hasNext()) {
```

Comment:
If this is still needed after addressing the above comment, can we use `Preconditions.checkState` instead? Also, `String.format` for the error message.